KAFKA-18015: add byDuration auto.offset.reset to Kafka Streams

mjsax opened the pull request:
Part of KIP-1106.

Adds support for "by_duration" and "none" reset strategy to the Kafka Streams runtime.

mjsax commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/OffsetResetStrategy.java:
```diff
@@ -0,0 +1,23 @@
+ */
+package org.apache.kafka.streams.kstream.internals;
+
+public enum OffsetResetStrategy {
```
Maybe we could also re-use `AutoOffsetResetStrategy.StrategyType`?

The weird thing is, that I realized that the exiting `Topology.AutoOffsetReset` does only have EARLIEST and LATEST, but not NONE... NONE is most likely not super useful and nobody asked for it yet, so maybe that's ok. It's just a question of re-using ``AutoOffsetResetStrategy.StrategyType` is the right thing, if we don't allow for NONE -- or maybe it does not matter.

It's also all internal code, we we can change it at any point anyway.

bbejeck replied:
> Topology.AutoOffsetReset does only have EARLIEST and LATEST, but not NONE

I'm guessing that was an oversight.  But overall I'm in favor of adding a `NONE` option as I prefer to give users the option to be explicit.  I also think re-using `AutoOffsetResetStrategy.StategyType` is a good thing to do, as we should align with clients behavior.

mjsax replied:
\cc @omkreddy for another small KIP update

mjsax replied:
Will add this to https://github.com/apache/kafka/pull/17973

ableegoldman replied:
just to clarify, the conclusion is that we'll add a `NONE` option to `Topology.AutoOffsetReset` as part of the KIP, and then use `AutoOffsetResetStrategy.StrategyType` internally?

mjsax replied:
The KIP deprecates `Topology.AutoOffsetReset` in favor of `AutoOffsetReset` class -- we will add a static builder `none()` to this new class for public usage.

And yes, we will use `AutoOffsetResetStrategy.StrategyType` internally (instead of adding our own `OffsetResetStrategy` enum) -- added this to the other PR already.

ableegoldman replied:
ah, thanks for clarifying. yeah that seems like the right call

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/Topology.java:
```diff
@@ -326,7 +327,7 @@ public synchronized Topology addSource(final AutoOffsetReset offsetReset,
                                            final Deserializer<?> valueDeserializer,
                                            final String... topics) {
-        internalTopologyBuilder.addSource(offsetReset, name, null, keyDeserializer, valueDeserializer, topics);
+        internalTopologyBuilder.addSource(offsetReset == null ? null : new AutoOffsetResetInternal(offsetReset), name, null, keyDeserializer, valueDeserializer, topics);
```
maybe refactor `offsetReset == null ? null : new AutoOffsetResetInternal(offsetReset)` into a method since it's used in several places.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/AutoOffsetResetInternal.java:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    public Duration duration() {
+        return null;
```
why a plain `null` here?

mjsax replied:
Not use yet, as the PR is incomplete. Working on this right now.

bbejeck reviewed (commented):
I made an initial pass.

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/AutoOffsetReset.java:
```diff
@@ -0,0 +1,21 @@
+ */
+package org.apache.kafka.streams;
+
+public class AutoOffsetReset {
```
nit: kind of confusing since we have an enum of the same name in the `Topology` class (not to mention it's empty -- guessing it will be filled in with/after the other PR?)

mjsax replied:
What do you mean by "it's empty" -- the KIP does actually deprecate `Topology.AutoOffsetReset` in favor of this new class (we need a class, as we need to not just track the strategy itself, but for `by_duration` the additional reset duration

cf the other PR

ableegoldman replied:
sorry, should have looked at the KIP and the other PR first 😞 shame on me

ableegoldman reviewed (commented):
Didn't give a super thorough review since it's not done yet but the general idea makes sense to me

bbejeck commented on streams/integration-tests/src/test/java/org/apache/kafka/streams/integration/FineGrainedAutoResetIntegrationTest.java:
```diff
@@ -251,14 +263,108 @@ private void commitInvalidOffsets() {
+        final Properties producerConfig = TestUtils.producerConfig(CLUSTER.bootstrapServers(), StringSerializer.class, StringSerializer.class);
+
+        for (int i = 0; i < 10; ++i) {
+            mockTime.sleep(Duration.ofHours(1L).toMillis());
```
How does this work? Unless I'm missing something, this looks to me like it should pause the test for an hour.

mjsax replied:
This is mock time, so there is no actual sleep. This test run just few seconds locally.

I used "an hour" because there is still some jitter (even with mock time), to ensure the test is stable.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java:
```diff
@@ -144,7 +144,7 @@ public synchronized <K, V> KStream<K, V> stream(final String topic,
      */
     public synchronized <K, V> KStream<K, V> stream(final Collection<String> topics) {
-        return stream(topics, Consumed.with(null, null, null, null));
+        return stream(topics, Consumed.with(null, null));
```
You may have explained this before - but why this change? (is this from the other PR and due to the compiler and overloads?)

mjsax replied:
Yes, there is multiple `with` overload taking 4 parameters now, and the compiler cannot resolve it. So we could either insert a cast, or just or a different overload. As we only pass in `null` anyway, I pick the second option.

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -1344,19 +1360,30 @@ private <S extends StateStore> InternalTopicConfig createChangelogTopicConfig(fi
             return AutoOffsetResetStrategy.LATEST;
+        } else if (maybeDecorateInternalSourceTopics(durationResetTopics.keySet()).contains(topic)) {
+            return AutoOffsetResetStrategy.fromString("by_duration:" + durationResetTopics.get(topic).toString());
+        } else if (findDuration(topic, resetDuration)) {
```
I'm thinking this could be simplified (part 1, part 2 in next comment below)

```
 public AutoOffsetResetStrategy offsetResetStrategy(final String topic) {
       final Optional<Duration> duration;
       if (...) {
           ....
       } else if ((duration = findDuration(topic)).isPresent()) {
                 return AutoOffsetResetStrategy.fromString("by_duration:" + duration);
      }
    ...
```

bbejeck commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -1366,6 +1393,19 @@ public AutoOffsetResetStrategy offsetResetStrategy(final String topic) {
         }
     }
 
+    private boolean findDuration(final String topic, final Set<Duration> resetDuration) {
```
Part 2 of simplification

```
private Optional<Duration> findDuration(final String topic) {
      return  durationResetPatterns.entrySet().stream()
            .filter(e -> e.getKey().matcher(topic).matches())
            .map(Map.Entry::getValue)
                .findAny();
    }
```

mjsax replied:
I don't link the `findAny()` part, as I wand to verify that we find not more than one item. But overall a good idea and I will adopt it.

bbejeck commented on streams/src/test/java/org/apache/kafka/streams/AutoOffsetResetTest.java:
```diff
@@ -0,0 +1,89 @@
+
+    @Test
+    void earliestShouldReturnAnEmptyDuration() {
+        final AutoOffsetResetInternal earliest = new AutoOffsetResetInternal(AutoOffsetReset.earliest());
```
nit: these 3 tests have the name `xShouldReturnAnEmptyDuration()` but the actual behavior is to throw a `NoSuchElementException` (which is expected by calling `get` on an empty `Optional`) - it's a little confusing it made my think it would return `Optional.empty()`

bbejeck reviewed (approved):
Overall LGTM - I left a few couple of comments you could address if you agree.

mjsax committed:
KAFKA-18015: add byDuration auto.offset.reset to Kafka Streams

Part of KIP-1106.

Adds support for "by_duration" and "none" reset strategy
to the Kafka Streams runtime.

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/internals/AutoOffsetResetInternal.java:
```diff
@@ -31,7 +30,9 @@ public AutoOffsetResetInternal(final AutoOffsetReset autoOffsetReset) {
-    public Optional<Duration> duration() {
-        return duration;
+
+    @SuppressWarnings("all")
```
...all?? 🧯

mjsax replied:
Yeah... not happy about this either... Could not find a dedicate suppress for this warning when accessing an Optional w/o `isPresent()` check... (What we do on purpose here...)

Given that the method is just a one-liner, I thought it should be ok?

ableegoldman replied:
maybe just leave a comment to explain why?

mjsax replied:
Works for me -- let's see what other things are there -- might just do a follow up PR to address this if ok, to avoid retriggering CI

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -1347,19 +1360,30 @@ private <S extends StateStore> InternalTopicConfig createChangelogTopicConfig(fi
     public boolean hasOffsetResetOverrides() {
-        return !(earliestResetTopics.isEmpty() && earliestResetPatterns.isEmpty()
-            && latestResetTopics.isEmpty() && latestResetPatterns.isEmpty());
+        return noneResetTopics.size() + noneResetPatterns.size()
```
out of curiosity why move away from the `isEmpty && isEmpty` pattern?

mjsax replied:
This is some checkstyle nonsense -- build failed with "too complex condition" warning/error after I added `[none|duration]Reset[Topic|Pattern].isEmpty()`. -- Instead of adding a suppression for it, I just rewrote it to make checkstyle happy.

ableegoldman replied:
🤷‍♀️

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -1347,19 +1360,30 @@ private <S extends StateStore> InternalTopicConfig createChangelogTopicConfig(fi
 
     public AutoOffsetResetStrategy offsetResetStrategy(final String topic) {
-        if (maybeDecorateInternalSourceTopics(earliestResetTopics).contains(topic) ||
+        Optional<Duration> resetDuration;
```
github/checkstyle is saying this should be final

ableegoldman replied:
I know you already fixed this, but I'd left the review comment before you pushed the last commit to fix spotBugs, and wasn't sure how to remove this from the staged review comments 🤷‍♀️

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/InternalTopologyBuilder.java:
```diff
@@ -1347,19 +1360,30 @@ private <S extends StateStore> InternalTopicConfig createChangelogTopicConfig(fi
+            return AutoOffsetResetStrategy.fromString("by_duration:" + resetDuration.get());
         } else if (containsTopic(topic)) {
-            return AutoOffsetResetStrategy.NONE;
+            return null;
```
why change this from `NONE` to `null`?

IIUC this is the reason you then had to change the `TopologyMetadata#offsetResetStrategy`'s return type to `Optional<AutoOffsetResetStrategy>`, which imo made [this](https://github.com/apache/kafka/pull/18115/files#diff-76f629d0df8bd30b2593cbcf4a2dc80de3167ebf55ef8b5558e6e6285a057496R1304) part of the StreamThread code much harder to follow than necessary. Don't like mixing Optional.empty and null

mjsax replied:
We need one more "condition" now that we support "none"... We return NONE above now.

This NONE was kinda weird -- we used it to indicate "topic has no per-topic reset policy -> use the one from StreamsConfig" -- what was semantically incorrect to begin with (but it was internal code, so it did not really matter).

Does this make sense?

ableegoldman replied:
ah, that makes sense. thanks

mjsax committed:
spotbug

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1291,29 +1293,77 @@ private void resetOffsets(final Set<TopicPartition> partitions, final Exception
+            final Optional<AutoOffsetResetStrategy> offsetResetStrategy = topologyMetadata.offsetResetStrategy(partition.topic());
 
             // This may be null if the task we are currently processing was apart of a named topology that was just removed.
             // TODO KAFKA-13713: keep the StreamThreads and TopologyMetadata view of named topologies in sync until final thread has acked
```
can we update this TODO to make it about removing the null check once we remove named topologies

mjsax replied:
Sure, I can update it to be more explicit -- I do read it as "remove null check after NT are removed" already though

mjsax replied:
Could do this in a follow up PR, too, to avoid waiting for CI again...

ableegoldman replied:
was less concerned about us misinterpreting the TODO, more that it just seems weird to have a TODO in the code for a ticket we never plan to do (though that reminds me, I should go close that ticket)

ableegoldman commented on streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamThread.java:
```diff
@@ -1326,6 +1376,50 @@ private void resetOffsets(final Set<TopicPartition> partitions, final Exception
+                    }
+                } catch (final TimeoutException timeoutException) {
+                    for (final TopicPartition partition : seekByDuration.keySet()) {
+                        final Task task = taskManager.getActiveTask(partition);
+                        task.maybeInitTaskTimeoutOrThrow(now, timeoutException);
+                        stateUpdater.add(task);
```
nit: i don't love making `#getActiveTask` non-private like this, can we just wrap these calls in a new TaskManager API

mjsax replied:
Happy to take a look -- can we do this in a follow up PR so we can merge this PR on-time?

I assume it could imply larger refactoring what would not belong into this PR.

ableegoldman replied:
fine with me

ableegoldman reviewed (approved):
LGTM -- feel free to merge as-is since the build passed, and address my remaining feedback in a follow up
